pkg/asset/agent/gencrypto/authconfig_test.go

```diff
@@ -26,37 +75,190 @@ func TestAuthConfig_Generate(t *testing.T) {
 
-			claims, err = ParseToken(authConfigAsset.UserAuthToken)
-			assert.NoError(t, err)
-			persona, ok = claims["auth_scheme"].(string)
-			assert.Equal(t, persona, userPersona)
-			assert.Equal(t, ok, true)
+				// Verify each token is signed with the correct persona
+				assertTokenPersona(t, authConfigAsset.AgentAuthToken, agentPersona)
+				assertTokenPersona(t, authConfigAsset.UserAuthToken, userPersona)
+				assertTokenPersona(t, authConfigAsset.WatcherAuthToken, watcherPersona)
+			}
 
-			claims, err = ParseToken(authConfigAsset.WatcherAuthToken)
-			assert.NoError(t, err)
-			persona, ok = claims["auth_scheme"].(string)
-			assert.Equal(t, persona, watcherPersona)
-			assert.Equal(t, ok, true)
+			if tc.workflow == workflow.AgentWorkflowTypeAddNodes {
+				// Retrieve the cluster secret and verify there are no errors
+				clusterSecret, err := fakeClient.CoreV1().Secrets(authTokenSecretNamespace).Get(context.Background(), authTokenSecretName, metav1.GetOptions{})
+				assert.NoError(t, err)
+
+				var secretTokens map[string]string
+				if tc.tokenExpired {
+					// Verify secret annotations are updated when tokens are expired.
+					assertAnnotations(t, clusterSecret, false)
+
+					if tc.expectedNumberOfAuthTokens == 3 {
+						// Ensure all 3 tokens (agentAuthToken, userAuthToken, watcherAuthToken) are present for OCP 4.18+.
+						secretTokens = verifyAllTokensPresent(t, clusterSecret.Data, agentAuthKey, userAuthKey, watcherAuthKey)
+						// Verify expired tokens in the cluster are replaced with tokens from the asset store.
+						assertTokensConsistency(t, assetStore3Tokens, secretTokens, false)
+					} else {
+						// Ensure only agentAuthToken is present in the secret for OCP 4.17.
+						secretTokens = verifyAllTokensPresent(t, clusterSecret.Data, agentAuthKey)
+						// Verify expired tokens in the cluster are replaced with tokens from the asset store.
+						assertTokensConsistency(t, assetStore1Token, secretTokens, false)
+					}
+				} else {
+					// Verify secret annotations are unchanged when tokens are valid.
+					assertAnnotations(t, clusterSecret, true)
+
+					if tc.expectedNumberOfAuthTokens == 3 {
+						// Ensure all 3 tokens (agentAuthToken, userAuthToken, watcherAuthToken) are present for OCP 4.18+.
+						secretTokens = verifyAllTokensPresent(t, clusterSecret.Data, agentAuthKey, userAuthKey, watcherAuthKey)
+						// Verify valid tokens in the cluster match those in the asset store.
+						assertTokensConsistency(t, assetStore3Tokens, secretTokens, true)
+					} else {
+						// Ensure only agentAuthToken is present in the secret for OCP 4.17.
+						secretTokens = verifyAllTokensPresent(t, clusterSecret.Data, agentAuthKey)
+						// Verify valid tokens in the cluster match those in the asset store.
+						assertTokensConsistency(t, assetStore1Token, secretTokens, true)
+					}
+				}
+			}
 		})
 	}
 }
+
+func secretWithOnly1Token(expiry time.Time) func(t *testing.T) runtime.Object {
```

Comment:
I think the secretWithOnly1Token and secretwith3Tokens functions are circumventing the goal of the unit tests. AuthConfig.Generate and the unit tests should be calling the same sets of functions to create the Secrets that stores the tokens. Otherwise the unit test isn't testing the actual code.